src/modules/roc_sndio/target_pulseaudio/roc_sndio/pulseaudio_backend.cpp

```diff
@@ -17,6 +17,22 @@
 namespace roc {
 namespace sndio {
 
+namespace {
+
+bool is_pulse_running() {
+    pa_context* context =
+        pa_context_new(pa_mainloop_get_api(pa_mainloop_new()), "ROC Daemon Check");
+
+    if (int err = pa_context_connect(context, NULL, PA_CONTEXT_NOFLAGS, NULL)) {
```

Comment:
Several concerns here:

1. We call pa_mainloop_get_api(), but never call pa_mainloop_free(), it's a leak.

2. We don't check for return values of pa_mainloop_new(), pa_mainloop_get_api(), and pa_context_new(). If they return NULL, we'll crash. We should return false instead.

3. Is just calling pa_context_connect() really enough? Generally speaking, it's an asynchronous function, and I'd expect that we should run the main loop (currently we never start it) and wait for state callback, and then check whether connection failed or succeeded. Or is there a reason why we don't need to wait callback in this specific case?

Reply:
It seems that correctly waiting for pa_context_connect() callback requires quite complicated code. However, we actually already have this code in PulseaudioSink implementation.

Maybe it would be wiser to get rid of probe() at all?

Currently, BackendDispatcher calls probe() in a loop until it succeeds, and then just call open_source() or open_sink() on the given backend. It seems we can easily replace it with a loop which calls open_source() or open_sink() until it succeeds. We just need to remove probe() and instead pass flags directly to open_source() and open_sink(). We also should move corresponding flag checks from probe to open, of course.

This approach will allow us to re-use existing pulseaudio logic, and also will avoid creating pulseaudio connection twice, which is nice.

Reply:
@gavv , thanks for your comments. Yes, you are right about pa_context_connect() being an async call. I overlooked that. Your suggestion might be wiser to implement. I will work on it soon.